public/static/app/perfdiscovery/PerfDiscoveryStateService.js

```diff
@@ -0,0 +1,168 @@
+mciModule.factory('PerfDiscoveryStateService', function(
+  $httpParamSerializer, $location
+) {
+  /* STATE ADAPTORS */
+
+  function sortToState(cols) {
+    return {
+      sort: _.reduce(cols, function(m, d) {
+        m[d.field] = d.sort
+        return m
+      }, {})
+    }
+  }
+
+  function onSortChanged(state) {
+    return function(grid, cols) {
+      applyState(state, sortToState(cols))
+    }
+  }
+
+  function filteringToState(cols) {
+    return {
+      filtering: _.reduce(cols, function(m, d) {
+        m[d.field] = d.filters[0].term
+        return m
+      }, {})
+    }
+  }
+
+  function onFilteringChanged(state, grid) {
+    return function() {
+      applyState(
+        state,
+        filteringToState(
+          _.filter(
+            grid.columns, function(col) { return col.filters[0].term }
+          )
+        )
+      )
+    }
+  }
+
+  /* SERIALIZATION */
+
+  function serializeFromVersion(state) {
+    return {from: state.from}
+  }
+
+  function serializeToVersion(state) {
+    return {to: state.to}
+  }
+
+  function serealizeSorting(state) {
+    return {
+      sort: _.map(state.sort, function(v, k) {
+        return (v.direction == 'desc' ? '-' : '') + k + v.priority
+      })
+    }
+  }
+
+  function serealizeFiltering(state) {
+    return _.reduce(state.filtering, function(m, v, k) {
+      m[k] = v
+      return m
+    }, {})
+  }
+
+  /* DESERIALIZATION */
+
+  function deserializeFromVersion(serialized) {
+    var from = serialized.from
+    return from ? {from: from} : {}
+  }
+
+  function deserializeToVersion(serialized) {
+    var to = serialized.to
+    return to ? {to: to} : {}
+  }
+
+  function deserializeSorting(serialized) {
+    var re = /(-?)(\w+)(\d)/
```

Comment:
Does this work correctly if a key has a number in it?

Reply:
No, it will work incorrectly. I don't think there are exists any examples/use cases of such fields.